### PR TITLE
Add example to use docker buildx via the custom builder

### PR DESCRIPTION
**Pull request by briandealwis**
**Description**
This example demonstrates how to use `docker buildx` as a Skaffold custom builder.

The one downside is that it requires setting up a `docker buildx` builder instance as a separate step.  Something like:

    docker buildx create --name skaffold-builder --platform linux/arm64,linux/amd64

We could create a more sophisticated build script that would examine the existing builder instances, but that's a bit more involved.

### CODE REVIEW

File: integration/examples/custom-buildx/main.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package main
+
+import (
+	"fmt"
+	"time"
+)
+
+func main() {
+	for {
+		fmt.Println("Hello world!")
```
Comment (tejal29) [inline integration/examples/custom-buildx/main.go:10]: we could change this line to run 
```suggestion
		fmt.Printf("hello multi platform, I am %s-%s\n", runtime.GOOS, runtime.GOARCH) 
```

File: integration/examples/custom-buildx/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.16-alpine as builder
+COPY main.go .
+# `skaffold debug` sets SKAFFOLD_GO_GCFLAGS to disable compiler optimizations
+ARG SKAFFOLD_GO_GCFLAGS
+RUN go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -o /app main.go
```
Comment (tejal29) [inline integration/examples/custom-buildx/Dockerfile:5]: should we use `TARGETARCH` and `TARGETOS`  args here and pass them to `go build`
```suggestion
FROM --platform=$BUILDPLATFORM golang:alpine AS build                                                                       
                                                                       
ARG TARGETOS                                                                                                                
ARG TARGETARCH                                                                                                                                                                                          
RUN GOOS=$TARGETOS GOARCH=$TARGETARCH  go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -o /app main.go 
```

File: integration/examples/custom-buildx/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+$ skaffold run --tail --default-repo <your repo>
+```
+
+With this command, Skaffold will build the artifact with `docker buildx`
+and deploy the application to Kubernetes.  You should be able to
+see *Hello, World!* printed every second in the Skaffold logs.
+
+If Skaffold fails with a message like the following, then Skaffold is
+attempting to load the multi-platform images into the Docker Daemon,
+is pushing the images
```
Comment (tejal29) [inline integration/examples/custom-buildx/README.md:69]: Not sure i understand what this means. 
Is there a way to recover from this error?

File: integration/examples/custom-buildx/skaffold.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  # Note that building for multiple platforms requires pushing to a registry,
+  # as the Docker Daemon cannot load multi-platform images.
+  local:
+    push: true
+  artifacts:
+  - image: skaffold-examples-buildx
+    custom:
+      # This buildCommand depends on a `docker buildx` builder to have been created
+      # with the name `skaffold-builder`:
+      #   $ docker buildx create --name skaffold-builder --platform linux/amd64,linux/arm64
```
Comment (tejal29) [inline integration/examples/custom-buildx/skaffold.yaml:14]: do we also need to add the `docker buildx  use skaffold-builder` command. 
or add the `--use`  flag to `docker buildx create --name skaffold-builder --platform linux/amd64,linux/arm64` command.

File: integration/examples/custom-buildx/README.md (hunk 3 above)
Comment (briandealwis) [reply integration/examples/custom-buildx/README.md:69]: Oops, it looks like a line was deleted.

File: integration/examples/custom-buildx/skaffold.yaml (hunk 4 above)
Comment (briandealwis) [reply integration/examples/custom-buildx/skaffold.yaml:14]: I have a slight preference to explicitly specifying the builder with `--builder` as the message if the user *hasn't* created a builder and continues to use the default docker builder is cryptic:

    error: multiple platforms feature is currently not supported for docker driver. Please switch to a different driver (eg. "docker buildx create --use")


Whereas explicitly specifying `--builder xxx` where `xxx` doesn't exist fails with a slightly more coherent message:

    error: open /Users/bdealwis/.docker/buildx/instances/foo: no such file or directory

File: examples/custom-buildx/README.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
+### Example: use the custom builder with `docker buildx`
+
+[Docker Buildx](https://github.com/docker/buildx#buildx) is an
+experimental feature for building container images for multiple
+platforms.
+
+This example shows how `docker buildx` can be used as a
+Skaffold _custom builder_ to create container images for
+for two different platforms: `linux/arm64` and `linux/amd64`.
```
Comment (PriyaModali) [inline examples/custom-buildx/README.md:9]: Remove duplicate `for`

File: examples/custom-buildx/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+Take a look at the [skaffold.yaml](skaffold.yaml), which uses a
+_custom builder_ to invoke `docker buildx` to containerize source
+code for two platforms.
+For more information about custom builders, see the Skaffold custom
+builder [documentation](https://skaffold.dev/docs/how-tos/builders/#custom-build-script-run-locally).
+Note that Skaffold builders are different from `docker buildx` builders.
+
+Next look at the Kubernetes [pod descriptor](k8s/pod.yaml) and notice
+the use of _node affinities_ to instruct Kubernetes to scheduler the workload
```
Comment (PriyaModali) [inline examples/custom-buildx/README.md:56]: Replace `scheduler` with `schedule`.

**Review by PriyaModali — DISMISSED** [review]

**Comment by briandealwis** [comment]
PTAL @PriyaModali

File: examples/custom-buildx/README.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  - When `true`, the result is to be pushed to a registry, and all
+    registries support multi-platform images. 
+  - When `false`, the result is to be loaded to the Docker Daemon. 
+    But Buildx does not support loading multi-platform images to
+    the Docker Daemon, and so this example builds a single container
+    image for the local platform.
+
+But Skaffold is unaware that the build result differs based on `$PUSH_IMAGE`.
+So on a local-build, Skaffold will cache the local-platform artifact,
+and that artifact will be used that for subsequent deployments even if
```
Comment (tejal29) [inline examples/custom-buildx/README.md:118]:
```suggestion
and that artifact will be used that for subsequent deployments even if pushed
```

File: examples/custom-buildx/README.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+to a remote registry (assuming the source is unchanged).  To avoid
+this scenario, disable Skaffold's artifact caching when the result
```
Comment (tejal29) [inline examples/custom-buildx/README.md:120]: This makes me wonder, if users used `mult-arch` images in dev loop but in `build` command for ci-cd purpose when they want to build a multi-arch images.

If you could document something along the lines at the beginning, 
```
Using `buildx` to build for multiple platforms has some subtle
interactions with Skaffold's artifact caching.
We don't recommend using `buildx` in skaffold dev mode. Using `buildx` is better suited for building and pushing multi-arch images in your CI release pipeline.

```

File: examples/custom-buildx/README.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+depending on the `$PUSH_IMAGE` flag. 
+
+  - When `true`, the result is to be pushed to a registry, and all
+    registries support multi-platform images. 
+  - When `false`, the result is to be loaded to the Docker Daemon. 
+    But Buildx does not support loading multi-platform images to
+    the Docker Daemon, and so this example builds a single container
+    image for the local platform.
+
+But Skaffold is unaware that the build result differs based on `$PUSH_IMAGE`.
```
Comment (tejal29) [inline examples/custom-buildx/README.md:116]:
```suggestion 
Skaffold is unaware that the build result differs based on `$PUSH_IMAGE`.
```

**Review by tejal29 — CHANGES_REQUESTED** [review]

File: examples/custom-buildx/README.md (hunk 9 above)
Comment (briandealwis) [reply examples/custom-buildx/README.md:116]: The "But" is helpful here as it is contrary to expectations.

File: examples/custom-buildx/README.md (hunk 8 above)
Comment (briandealwis) [reply examples/custom-buildx/README.md:120]: The dev+multi-platform is an excellent observation and one that we'll take it into account with multi-platform support in Skaffold. (DD to come RSN.)

I'll add a note about `dev` use and suggest to use command-activated profiles.  This example is hairy enough! :-)

File: examples/custom-buildx/README.md (hunk 9 above)
Comment (tejal29) [reply examples/custom-buildx/README.md:116]: they were too many buts. There was a but in the sentence above too. I am not an expert on english **but* in my mind  all sentences after the first "But" inherit the "but" clause?

**Review by tejal29 — APPROVED** [review]

File: examples/custom-buildx/README.md (hunk 8 above)
Comment (tejal29) [reply examples/custom-buildx/README.md:120]: yeah i didn't say explicitly *but* in my mind I meant we can keep the example as is.

**Pull request merged**